core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -180,6 +181,11 @@ class KafkaApis(val requestChannel: RequestChannel,
         case ApiKeys.DESCRIBE_CLIENT_QUOTAS => handleDescribeClientQuotasRequest(request)
         case ApiKeys.ALTER_CLIENT_QUOTAS => handleAlterClientQuotasRequest(request)
       }
+
+      // try to complete delayed action. In order to avoid conflicting locking, the actions to complete delayed requests
+      // are kept in a queue. We add the logic to check the ReplicaManager queue at the end of KafkaApis.handle() and the
+      // expiration thread for certain delayed operations (e.g. DelayedJoin)
+      actionQueue.tryCompleteAction()
```

Comment:
Even if we hit an exception in handleXXX(), it would still be useful to complete the actionQueue.

Reply:
Completing delayed actions may cause exception. Should exception be swallowed and log if we move the completion to the final block?

Reply:
Yes, if actionQueue.tryCompleteAction() throws an exception, we can just catch it and log a warning in finally since the response has been sent by then.